design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,344 @@
+
+The cluster of KSQl server will be able to serve requests even when the active is down. This mitigates 
+a large current gap in deploying KSQL pull queries for mission critical use-cases, by significantly 
+reducing failure rate of pull queries during server failures.
+
+## Public APIS
+Add configuration parameter to the `WITH` clause of CTAS statements to specify `acceptable.offset.lag`. Also, add property to the json request to specify the above property.
+
+## Design
+
+The problem at hand is how to determine if a server is down and where to route a request when it is 
+down. Every KSQL server must have available the information of what other KSQL servers exist in the 
+cluster, their status and their lag per topic partition. There are three components to this: 
+Healthchecking, Lag reporting and Lag-aware routing.  
+
+### Failure detection/Healthcheck
+
+Failure detection/ healthcheck is implemented via a periodic hearbeat. KSQL servers either broadcast 
+their heartbeat (N^2 interconnections with N KSQL servers) or we implement a gossip protocol. In 
+the initial implementation, we will use the REST API to send heartbeats leveraging the N^2 mesh that 
+already exists between KSQL servers. Hence, we will implement a new REST endpoint, `/heartbeat` that 
+will register the heartbeats a server receives from other servers. 
+
+Cluster membership is determined using the information provided from the 
+[Kafka Streams](https://kafka.apache.org/20/javadoc/org/apache/kafka/streams/KafkaStreams.html)
+instance and specifically [`StreamsMetadata`](https://kafka.apache.org/20/javadoc/org/apache/kafka/streams/state/StreamsMetadata.html) 
+from which we can obtain the information of the host and port of the instance. A server periodically
+polls for all currently running KS instances (local and remote) and updates the list of remote servers
+it has seen. If no KS streams are currently running, cluster membership is not performed. Moreover,
+this policy depends on the current design where a new KSQL server replays every command in the 
+command topic and hence runs every query.
+
+We want to guarantee 99% uptime. This means that in a 5 minute window, it must take no longer than 
+2 seconds to determine the active server is down and to forward the request to one of the standbys. 
+The heartbeats must be light-weight so that we can send multiple heartbeats per second which will 
+provide us with more datapoints required to implement a well-informed policy for determining when a 
+server is up and when down.
+
+Configuration parameters:
+1) Heartbeat INTERVAL (e.g send heartbeat every 200 ms)
+2) Heartbeat WINDOW (e.g every 2 seconds, count the missed/received heartbeats and determine if server is down/up)
+3) MISSED_THRESHOLD: How many heartbeats in a row constitute a node as down (e.g. 3 missed heartbeats = server is down)
+4) RECEIVED_THRESHOLD: How many heartbeats in a row constitute a node as up (e.g. 2 received heartbeats = server is up)
```

Comment:
For the example you describe, the server will be marked as down for that window as the heartbeats will be considered as missed. I think this is OK as in the next window, if it heartbeats enough, it will be marked as up again. The window is a sliding window and it is needed for counting the missed/received heartbeats from `now` (when the counting occurs) until `now-window`.

I haven't thought about servers having different `INTERVALs`. Wouldn't we want to disallow this?  If we do want this information, it makes more sense to get it when a server is first discovered  and not in every heartbeat message due to size overhead since this is configured at startup of the server and should not change. Maybe I need to add a new endpoint for getting configuration parameters of discovered servers?

Reply:
@big-andy-coates I added this text:

> Description of how these configuration parameters are used:
> 
> Every server sends a heartbeat every SEND interval. Every server processes its received heartbeats via the 
> `decide-status` process every CHECK interval. The `decide-status` process considers only heartbeats that are 
> received from `windowStart = now - WINDOW` to `windowEnd = now`. Heartbeats that were received before `windowStart`
> are expunged. The `decide-status` process counts the number of missed and received heartbeats in one window by 
> checking whether there was a heartbeat received every SEND interval starting from `windowStart`. For example, from `windowStart=0` and `SEND=100` it will check if there was a heartbeat received at timestamp 100, 200, 300 etc. until `windowEnd`. If there is a timestamp for which no heartbeat was received, the process increases the missed count. If there are more than MISSED_THRESHOLD heartbeats missed in a row, then a server is marked as down.

Does this explain why we need all the parameters?

Reply:
Added ticket #4261

Reply:
If have similar concerns as @big-andy-coates  (https://github.com/confluentinc/ksql/pull/4022#discussion_r364338037).

For the miss-configuration: I would recommend to "compare server configs" with a different mechanism and not piggyback it into the actual heartbeats. It seems to be two different concerns.